File: client-sdks/runestone.rb
# Runestone OpenAI-Compatible API Client for Ruby

require 'net/http'
require 'json'
require 'uri'

module Runestone
  class Client
    attr_reader :api_key, :base_url

    def initialize(api_key:, base_url: 'http://localhost:4001/v1')
      @api_key = api_key
      @base_url = base_url
    end

    def chat_completion(model:, messages:, temperature: nil, max_tokens: nil, stream: false)
      request_body = {
        model: model,
        messages: messages,
        temperature: temperature,
        max_tokens: max_tokens,
        stream: stream
      }.compact

      post('/chat/completions', request_body)
    end

    def list_models
      get('/models')
    end

    def get_model(model_id)
      get("/models/#{model_id}")
    end

    def create_embeddings(model:, input:)
      post('/embeddings', { model: model, input: input })
    end

    private

    def post(endpoint, body)
      uri = URI("#{@base_url}#{endpoint}")
      http = Net::HTTP.new(uri.host, uri.port)
      
      request = Net::HTTP::Post.new(uri)
      request['Authorization'] = "Bearer #{@api_key}"
      request['Content-Type'] = 'application/json'
      request.body = body.to_json

      response = http.request(request)
      handle_response(response)
    end

    def get(endpoint)
      uri = URI("#{@base_url}#{endpoint}")
      http = Net::HTTP.new(uri.host, uri.port)
      
      request = Net::HTTP::Get.new(uri)
      request['Authorization'] = "Bearer #{@api_key}"

      response = http.request(request)
      handle_response(response)
    end

    def handle_response(response)
      case response.code.to_i
      when 200
        JSON.parse(response.body)
      when 401
        raise AuthenticationError, "Invalid API key"
      when 429
        raise RateLimitError, "Rate limit exceeded"
      else
        raise APIError, "API error: #{response.body}"
      end
    end
  end

  class APIError < StandardError; end
  class AuthenticationError < APIError; end
  class RateLimitError < APIError; end
end

# Example usage:
# client = Runestone::Client.new(api_key: 'sk-test-key')
# response = client.chat_completion(
#   model: 'gpt-4o-mini',
#   messages: [{ role: 'user', content: 'Hello!' }]
# )

File: client-sdks/runestone.go
// Package runestone provides a client for the Runestone OpenAI-compatible API
package runestone

import (
    "bytes"
    "encoding/json"
    "fmt"
    "io"
    "net/http"
)

// Client represents a Runestone API client
type Client struct {
    APIKey  string
    BaseURL string
    http    *http.Client
}

// NewClient creates a new Runestone client
func NewClient(apiKey string) *Client {
    return &Client{
        APIKey:  apiKey,
        BaseURL: "http://localhost:4001/v1",
        http:    &http.Client{},
    }
}

// ChatMessage represents a chat message
type ChatMessage struct {
    Role    string `json:"role"`
    Content string `json:"content"`
}

// ChatCompletionRequest represents a chat completion request
type ChatCompletionRequest struct {
    Model       string        `json:"model"`
    Messages    []ChatMessage `json:"messages"`
    Temperature float64       `json:"temperature,omitempty"`
    MaxTokens   int           `json:"max_tokens,omitempty"`
    Stream      bool          `json:"stream,omitempty"`
}

// ChatCompletionResponse represents a chat completion response
type ChatCompletionResponse struct {
    ID      string `json:"id"`
    Object  string `json:"object"`
    Created int64  `json:"created"`
    Model   string `json:"model"`
    Choices []struct {
        Index   int         `json:"index"`
        Message ChatMessage `json:"message"`
    } `json:"choices"`
    Usage struct {
        PromptTokens     int `json:"prompt_tokens"`
        CompletionTokens int `json:"completion_tokens"`
        TotalTokens      int `json:"total_tokens"`
    } `json:"usage"`
}

// Model represents an AI model
type Model struct {
    ID      string `json:"id"`
    Object  string `json:"object"`
    Created int64  `json:"created"`
    OwnedBy string `json:"owned_by"`
}

// CreateChatCompletion creates a chat completion
func (c *Client) CreateChatCompletion(req ChatCompletionRequest) (*ChatCompletionResponse, error) {
    data, err := json.Marshal(req)
    if err != nil {
        return nil, err
    }

    httpReq, err := http.NewRequest("POST", c.BaseURL+"/chat/completions", bytes.NewBuffer(data))
    if err != nil {
        return nil, err
    }

    httpReq.Header.Set("Authorization", "Bearer "+c.APIKey)
    httpReq.Header.Set("Content-Type", "application/json")

    resp, err := c.http.Do(httpReq)
    if err != nil {
        return nil, err
    }
    defer resp.Body.Close()

    if resp.StatusCode != http.StatusOK {
        body, _ := io.ReadAll(resp.Body)
        return nil, fmt.Errorf("API error: %s", string(body))
    }

    var result ChatCompletionResponse
    if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
        return nil, err
    }

    return &result, nil
}

// ListModels lists available models
func (c *Client) ListModels() ([]Model, error) {
    req, err := http.NewRequest("GET", c.BaseURL+"/models", nil)
    if err != nil {
        return nil, err
    }

    req.Header.Set("Authorization", "Bearer "+c.APIKey)

    resp, err := c.http.Do(req)
    if err != nil {
        return nil, err
    }
    defer resp.Body.Close()

    var result struct {
        Data []Model `json:"data"`
    }
    if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
        return nil, err
    }

    return result.Data, nil
}
